proposals/0000-abstract-q.rst

```diff
@@ -0,0 +1,191 @@
+
+By separating out the internal interface, we can avoid this tight coupling.
+``reifyCore`` can be added as a field to ``MetaHandlers``. ``Quasi`` would then live purely in ``template-haskell``.
+Old versions of ``template-haskell`` would still be compatible with the new version of GHC, as they can simply ignore the new field.
+A new version of ``template-haskell`` can add the corresponding method to ``Quasi`` and use the field from ``MetaHandlers`` to implement it.
+
+``Q`` is wired-in to GHC. This fixes a single definition of it for each version of GHC. Since it is implemented in terms of ``Quasi``, it is also fixed.
+By breaking this tight coupling, we allow ``template-haskell``\'s interface to potentially be compatible with a greater range of GHC versions and to evolve independently of it.
+
+Our new definition of ``Q`` is also easier to optimise for the compiler, since it uses a known ``Monad``. Though this is a minor benefit, since the runtime performance of splices is rarely an issue. It also has potential to interface well with the ``bluefin``/``effectful`` family of effects libraries, which are implemented in terms of ``IO``, and could not express Template Haskell effects previously.
+
+Proposed Change Specification
+-----------------------------
+
+No changes are being proposed to the language.
+
+Proposed Library Change Specification
+-------------------------------------
+
+This proposal requires making changes to the interface of ``template-haskell``.
+In this section we will purely focus on the changes to the interface.
+We will return to the implementation details in `Implementation plan <#7implementation-plan>`_
+
+The interface of ``Language.Haskell.TH.Syntax`` (and ``Language.Haskell.TH``) will change from::
+
+ -- Note: these is defined in ghc-internal:GHC.Internal.TH.Syntax
+ -- and only re-exported from template-haskell. These are wired-in definitions of GHC.
+
+ newtype Q a = Q { unQ :: forall m. Quasi m => m a }
+
+ class (MonadIO m, MonadFail m) => Quasi m where
+  qNewName :: String -> m Name
+  qRecover :: m a -> m a -> m a
+  qReport  :: Bool -> String -> m ()
+  qReify   :: Name -> m Info
+  ... and so on
+
+to::
+
+ -- Note: Q is defined in ghc-internal:GHC.Internal.TH.Syntax
+ -- and only re-exported from template-haskell.
+ -- It is wired-in.
+ newtype Q a -- Q is abstract or opaque
+
+ -- Note: Quasi is now defined in template-haskell. It is no longer wired-in.
+
+ class (MonadIO m, MonadFail m) => Quasi m where
+  qRun     :: Q a -> m a -- New method
+  qNewName :: String -> m Name
+  qRecover :: m a -> m a -> m a
+  qReport  :: Bool -> String -> m ()
+  qReify   :: Name -> m Info
+  ... and so on
+  {-# MINIMAL qRun qRecover #-}
+
+``unQ`` and the ``Q`` constructor would no longer be exported from ``template-haskell``.
+This is a breaking change.
```

Comment:
Would it be feasible for `template-haskell` to use a pattern synonym to provide backwards-compatible definitions of (data constructor) `Q` and `unQ`? (Even if they are immediately marked as deprecated.)

Reply:
The types are:
```
unQ :: Q a -> forall m. Quasi m => m a
Q :: (forall m. Quasi m => m a) -> Q a
```

I think we might be able to do:
```
unQ' :: Q a -> (forall m. Quasi m => m a)
unQ' k = qRunQ k
```

But I don't think we can shim the other one AFAICT.

I think exposing a pattern synonym would be good even if it would only be in one direction. I'll update the proposal soon